firerunner/bins/controller/controller.rs

```diff
@@ -0,0 +1,145 @@
+        Controller {
+            inner: Arc::new(Inner {
+                active_functions: Default::default(),
+                warm_functions: Default::default(),
+                channels: Default::default(),
+                max_channel: AtomicUsize::new(3),
+                seccomp_level,
+                cmd_line,
+                kernel,
+                function_configs,
+            }),
+            vsock_closer: None
+        }
+    }
+
+    pub fn schedule(&mut self, req: request::Request) {
+        self.inner.schedule(req);
+    }
+
+    pub fn ignite(&mut self) -> Handle {
+        let (response_sender, response_receiver) = channel();
+
+        let (manager_handle, vsock_closer) = listener::RequestManager::new(self.inner.channels.clone(), response_sender).spawn();
+        self.vsock_closer = Some(vsock_closer);
+
+        let inner = self.inner.clone();
+        let response_handle = thread::spawn(move || {
+            for response in response_receiver.iter() {
+                inner.process_response(response);
+            }
+        });
+
+        Handle (vec![manager_handle, response_handle])
+    }
+
+    pub fn kill_all(&mut self) {
+        while self.inner.active_functions.lock().unwrap().len() > 0 {
+            thread::yield_now();
+        }
+        self.vsock_closer.take().map(|mut c| c.close()).unwrap().unwrap();
+        self.inner.warm_functions.lock().unwrap().clear();
+    }
+}
+
+impl Inner {
+    pub fn schedule(&self, req: request::Request) {
+        match self.active_functions.lock().unwrap().entry(req.function.clone()) {
+            Entry::Occupied(mut entry) => {
+                let (sender, outstanding, _app) = entry.get_mut();
+                *outstanding += 1;
+                sender.send(req).expect("sending request");
+            },
+            Entry::Vacant(entry) => {
+                if let Some((sender, app)) = self.warm_functions.lock().unwrap().remove(entry.key()) {
+                    sender.send(req).expect("sending request");
+                    entry.insert((sender, 1, app));
+                } else if let Some(config) = self.function_configs.get(entry.key()) {
+                    let cid = self.max_channel.fetch_add(1, Ordering::Relaxed) as u32;
+                    let (req_sender, req_receiver) = channel();
+                    self.channels.lock().expect("poisoned lock").insert(cid, (entry.key().clone(), req_receiver));
+                    let app = VmAppConfig {
+                        kernel: self.kernel.clone(),
+                        instance_id: config.name.clone(),
+                        rootfs: config.runtimefs,
+                        appfs: Some(config.appfs),
+                        cmd_line: self.cmd_line.clone(),
+                        seccomp_level: self.seccomp_level,
+                        vsock_cid: cid,
+                        cpu_share: config.vcpus,
```

Comment:
I assume `cpu_share` of a vm is equivalent to # vcpus of the vm.

Reply:
Not exactly, but for now that's fine. We actually want cpu_share to be based on the memory size, since many sizes of functions will have exactly one VCPU, but each should have a different share of the host CPU in the cgroup